File: DesignPatterns/01_GOF23/DP_examples/Go/02_Structural/Bridge/abstraction.go
package Bridge

// 抽象类
type Shape struct {
	draw DrawAPI
}

func NewShape(draw DrawAPI) *Shape {
	return &Shape{draw: draw}
}
func (s *Shape) Draw(color Color) {
	s.draw.Draw()
	s.draw.Fill(color)
}

type ShapeAdvance struct {
	Shape
	onDrawCompleted func()
}

func NewShapeAdvance(draw DrawAPI, onDrawCompleted func()) *ShapeAdvance {
	return &ShapeAdvance{
		Shape:           *NewShape(draw),
		onDrawCompleted: onDrawCompleted,
	}
}
func (sa *ShapeAdvance) Draw(color Color) {
	sa.Shape.Draw(color)
	if sa.onDrawCompleted != nil {
		sa.onDrawCompleted()
	}
}

File: DesignPatterns/01_GOF23/DP_examples/C/02_Structural/Bridge/abstraction.c
#include "abstraction.h"

// 形状绘制方法
static void Shape_draw(Shape* self, Color color) {
    self->draw->draw(self->draw);
    self->draw->fill(self->draw, color);
}

// 初始化形状
void Shape_init(Shape* shape, DrawAPI* draw) {
    shape->draw = draw;
    shape->draw_shape = Shape_draw;
}

// 高级形状绘制方法
static void ShapeAdvance_draw(ShapeAdvance* self, Color color) {
    self->base.draw_shape(&self->base, color);
    if (self->on_draw_completed) 
        self->on_draw_completed();
}
void ShapeAdvance_init(ShapeAdvance* shapeAdvance, DrawAPI* draw, DrawCompletedCallback on_draw_completed) {
    Shape_init(&shapeAdvance->base, draw);
    shapeAdvance->on_draw_completed = on_draw_completed;
    shapeAdvance->draw_shape = ShapeAdvance_draw;
}
